src/daft-core/src/array/ops/cast.rs

```diff
@@ -128,15 +132,10 @@ where
         }
     };
 
-    let new_field = Arc::new(Field::new(to_cast.name(), dtype.clone()));
+    let new_field = Field::new(to_cast.name(), dtype.clone());
 
-    if dtype.is_logical() {
-        with_match_daft_logical_types!(dtype, |$T| {
-            return Ok(LogicalArray::<$T>::from_arrow(new_field.as_ref(), result_arrow_physical_array)?.into_series())
-        })
-    }
-    with_match_arrow_daft_types!(dtype, |$T| {
-        Ok(DataArray::<$T>::from_arrow(new_field.as_ref(), result_arrow_physical_array)?.into_series())
+    with_match_daft_types!(dtype, |$T| {
```

Comment:
I was able to use `with_match_daft_types!` instead of:
```
if logical {
    with_match_daft_logical_types!(...)
} else {
    with_match_daft_physical_types!(...)
}
```
in quite a few places. This is because we can now make use of `$T::ArrayType` instead of conditionally creating  `DataArray`/`LogicalArray` depending on the type.